packages/codegen/templates/app01/middlewares/oidc.js

```diff
@@ -0,0 +1,191 @@
+                    await req.session.save()
+
+                    return res.redirect(nextUrl)
+                } else {
+                    await req.session.destroy()
+                    return res.status(403).send('ERROR: Employee was not found!')
+                }
+            } catch (error) {
+                await req.session.destroy()
+                return next(error)
+            }
+        })
+
+        // Endpoint to start session for third-party embedded page while opened as first-party
+        app.get('/grant-storage-access/:id', async (req, res, next) => {
+            req.session['sessionStarted'] = true
+            await req.session.save()
+            return next()
+        })
+
+        app.get('/oidc/auth', async (req, res, next) => {
+            const checks = { nonce: generators.nonce(), state: generators.state() }
+            req.session[oidcSessionKey] = checks
+
+            const nextUrl = get(req, 'query.next')
+            if (isString(nextUrl)) {
+                req.session[nextUrlKey] = nextUrl
+            }
+
+            await req.session.save()
+            try {
+                const redirectUrl = helper.getAuthorizationUrlWithParams(checks)
+                return res.redirect(redirectUrl)
+            } catch (err) {
+                await req.session.destroy()
+                return next(err)
+            }
+        })
+
+        app.get('/oidc/callback', async (req, res, next) => {
+            try {
+                const checks = req.session[oidcSessionKey]
+                const nextUrl = req.session[nextUrlKey] || '/'
+                const launchParams = req.session[launchParamsKey]
+
+                if (!isObject(checks) || !checks) {
+                    await req.session.destroy()
+                    return res.status(400).send('ERROR: Invalid nonce and state')
+                }
+
+                const { accessToken, refreshToken, tokenExpiresAt, userInfo } = await helper.completeAuth(req, checks)
+                const user = await createOrUpdateUser(keystone, userInfo)
+                const appToken = await keystone._sessionManager.startAuthedSession(req, {
+                    item: { id: user.id },
+                    list: keystone.lists['User'],
+                })
+
+                req.session[CONDO_ACCESS_TOKEN_KEY] = accessToken
+                req.session[CONDO_REFRESH_TOKEN_KEY] = refreshToken
+                req.session[APP_TOKEN_KEY] = appToken
+                req.session[launchParamsKey] = launchParams
+
+                delete req.session[oidcSessionKey]
+                await req.session.save()
+
+                return res.redirect(nextUrl)
+            } catch (err) {
+                await req.session.destroy()
+                return next(err)
+            }
+        })
+
+        app.set('trust proxy', true)
+
+        return app
+    }
+}
+
+module.exports = {
+    CondoOIDCMiddleware,
```

Comment:
may be we need some `open-condo/oidc` package?

Reply:
There are even a todo by @Alllex202 if Im not mistaken :-D

Reply:
If you decide to put this logic in a separate package, then keep in mind that each miniapp may have its own unique logic

Reply:
> If you decide to put this logic in a separate package, then keep in mind that each miniapp may have its own unique logic

why would you want to have unique oidc auth logic (especially for our home-cooked miniapps?)